### ISSUE TITLE
In a multi-project build, working directory is inconsistent between forked and non-forked run and test

### ISSUE THREAD

**Issue by brikis98** [sbt/sbt#1032]
If you have an [SBT multiproject build](http://www.scala-sbt.org/release/docs/Getting-Started/Multi-Project) with multiple play apps, the typical working pattern is something like this:

```
> cd ~/source/foo
> play
[snip]
> project foo-frontend
> run
[snip]
> CTRL + D
> project foo-backend
> run
```

In all of these cases, since the JVM was originally launched from `~/source/foo`, the working directory (system property `user.dir`) will be set to `~/source/foo`.

However, if you fork the JVM to run tests, the forked JVM will get its working directory from SBT's `forkOptions` which are defined in [Defaults.scala](https://github.com/sbt/sbt/blob/0.13/main/src/main/scala/sbt/Defaults.scala#L404) as:

``` scala
private[this] def forkOptions: Initialize[Task[ForkOptions]] = (baseDirectory, javaOptions, outputStrategy, envVars, javaHome, connectInput) map {
  (base, options, strategy, env, javaHomeDir, connectIn) =>

    ForkOptions(bootJars = Nil, javaHome = javaHomeDir, connectInput = connectIn, outputStrategy = strategy, runJVMOptions = options, workingDirectory = Some(base), envVars = env)
}
```

Note how `workingDirectory` is set to `Some(base)`, which will be different for each sub-project in a multiproject build: `~/source/foo/foo-frontend` and `~/source/foo/foo-backend`.

This inconsistency between working directories breaks relative path lookups, since most methods in `java.io.File`, such as `exists()`, rely on `user.dir`.

**Comment by brikis98** [comment]
BTW, our current workaround is to manually reset the `workingDirectory` at test time as follows:

``` scala
testGrouping in Test := {
  val original: Seq[Tests.Group] = (testGrouping in Test).value

  original.map { group =>
    val forkOptions = ForkOptions(
      bootJars = Nil,
      javaHome = javaHome.value,
      connectInput = connectInput.value,
      outputStrategy = outputStrategy.value,
      runJVMOptions = javaOptions.value,
      workingDirectory = Some(new File(System.getProperty("user.dir"))),
      envVars = envVars.value
    )

    group.copy(runPolicy = Tests.SubProcess(forkOptions))
  }
}
```

**Comment by huntc** [comment]
Thanks!

**Comment by jsuereth** [comment]
So, a few points:
1. `baseDirectory` represents the base of the current project.  For mutli-module builds, that means the base directory of that sub-project.   The "root" project happens to share its `base-directory` with the build's `base-directory.   I think`baseDirectory in ThisBuild` will always give you consistently the base directory.
2.  I believe sbt's fork in run would be consistent with forked tests. see: [sbt's run definition](https://github.com/sbt/sbt/blob/0.13/main/src/main/scala/sbt/Defaults.scala#L417-L422) where runner is defined in a current project.
3. When _not forking_ the current directory is wherever sbt was launched.
4. The `baseDirectory` of the build may have no correlation to the directory where sbt is launched. Likelihood of this is very low (have to be using source-remote-build feature), so we'll ignore this for making a decision.
5. AFAIK, it's impossible to fake "workingDirectory" in  a process.  i.e. `sys.props("user.dir") = baseDirectory.value` will not work.

Now, the question is whether forking or not forking should cause differences in working directories.   I think you're right, we should be consistent across all projects in a build.   Given these 5 facts, it seems the right path is to always use the current working directory when forking proceses.

However this "feels wrong" to me.  I think this is solely because of the JVM restriction in point 5.  Personally, I'd prefer that anytime I do "run" or "test" the working directory is the base of that project (or sub-project).  That feels consistent and right.   It may also be impractical.

@harrah what do you think?

**Comment by richdougherty** [comment]
I don't think there's a good solution to this, apart from forking processes all the time. In my opinion forked and non-forked tasks should see the same working directory. I realise using SBT's working directory for forked process will break a lot of things, but the other alternatives will break things too. :)

I'm thinking about whether we should change the Play API to resolve relative paths from a directory specified in a system property, rather than from the current process's working directory. It might make sense to resolve from the "user.dir" property (since it can be overridden), or maybe we should use another property, e.g. "play.user.dir". By resolving paths from a property we effectively allow runtime changes of working directory, so non-forked tasks can set this property to the current SBT baseDirectory. In Play we could set this system property automatically in our task definitions for test, run, etc.

**Comment by brikis98** [comment]
`java.io.File` uses `user.dir` internally for resolving relative paths in some methods, such as `exists()`. If `user.dir` doesn't have the right (that is, consistent) value, many things will break in an unfixable way.

**Comment by jroper** [comment]
My 2 cents:
- Best practice is to not rely on a particular working directory ever.  Always read files from the classpath, and if you need to write files, use File.createTmpFile to create temporary files.  If you need to read from a directory, pass the directory from your build to your tests via a system property.  If you need a directory to write to, do:
  
  ```
  val file = File.createTmpFile("foo", "")
  file.delete()
  file.mkdir()
  ```
- Not everyone will follow the above, so they are going to have working directory dependent builds.  In which case, don't do anything with the working directory, let the user manage their own pain.
- Possibly provide an option for forking that allows you to specify the working directory.

**Comment by harrah** [comment]
Agree with the first point.  That is definitely best practice for those willing to follow it.  Can you elaborate what you mean in the second point by "don't do anything with the working directory"?  Do you mean that the forked directory should be the same as `user.dir`?  For the third point, it exists.  At least, it looks like @brikis98 is doing the long version of `baseDirectory in Test := ...`.  (That short version probably isn't documented as I've realized now it was a side effect of some other work in 0.13.0.)

**Comment by brikis98** [comment]
@harrah Does `baseDirectory in Test := ...` affect the `user.dir` system property? 

@jroper Agree that absolute paths are best practice, but not always easy, and occasionally impossible (e.g. if you're using a 3rd party library and can't control it). IMO, the "default" should work as expected; I find it very surprising that code would behave differently simply because the JVM was forked.

**Comment by jroper** [comment]
Yes, I mean have the working directory of the forked process be the same as the working directory of the parent process.

**Comment by harrah** [comment]
@brikis98 of the forked jvm, yes.  It should do the same as what your code snippet does.

On the proposal to make the forked jvm the same as `user.dir` of the build process, that is fine, but note that `user.dir` is not necessarily the build root.  It is true that in many cases it is.  However, it will not be when a) using source dependencies or b) if sbt searched for the root of a build if started in a subdirectory.  I understand these are not the common case.

The original reason for this behavior was that it seemed better to at least make the forked `user.dir` reliable even if the non-forked couldn't be.

Should the change be made, there is the problem of compatibility- how to deprecate and migrate?  Thanks for the feedback so far.

**Comment by brikis98** [comment]
@harrah Thanks for the shorter workaround, we'll change to use that one.

I'd recommend following the principle of least surprise. If I run the _exact same tests_, once with `fork := true` and once with `fork := false`, I would be (or, more accurately, I _was_) very surprised that relative file paths changed behavior. Since the original value of `user.dir` cannot be changed - it's the directory where you launched SBT - I think the only option is the backwards incompatible change of propagating that value to all forked JVMs.

For migration, I'm not sure what process SBT follows, but I'd guess that if you go down this route, clear instructions in the migration guide would be the way forward. Of course, documenting that absolute paths are always preferred is a good idea.

**Comment by jsuereth** [comment]
Yeah, I think it sounds the most reasonable to make `fork := false` and `fork := true` be consistent, and we can't change the `fork := false` case safely.

Definitely will need a migration guide when this is fixed.

**Comment by nchammas** [comment]
Just for the record, I ran into this issue of the working directory changing based on whether JVMs are forked or not while trying to [parallelize tests for Apache Spark](https://issues.apache.org/jira/browse/SPARK-3431). Passing in [`Some(baseDirectory.value)`](https://github.com/playframework/playframework/issues/1834#issuecomment-30029558) explicitly to the forked JVMs seemed to fix the issue. Thanks to @brikis98 and @muymoo for documenting their workaround.

**Comment by frgomes** [comment]
As @jroper mentioned, there are best practices that should be followed. But, as @jroper also mentioned, not everyone follow best practices, in particular when you have large legacy code bases.

Not everybody has chance of starting a brand new, beautiful and organized code base from scratch. In such cases, the pattern is, in the vast majority of cases, picking files relative to the current working directory. Again, in the vast majority of cases, the current working directory is the directory the project is sitting on, where "project" is some uncertain thing, I mean: it can either be the root directory or the subproject working directory.

Whatever scenario you are from the discussion above, you (as SBT user) expect that SBT adopts some sort of consistent working directory.

It should not make any difference whether you fork or not.
It should not make any difference if you are `run`ning or `test`ing or whatever it is.

Given that the current directory is always consistent, that's fine.

**Comment by frgomes** [comment]
Opened https://github.com/sbt/sbt/issues/2244

**Comment by olafurpg** [comment]
Would it be a possible workaround to set `user.dir` in `javaOptions`?

```scala
    javaOptions += s"-Duser.dir=${baseDirectory.in(ThisBuild).value}",
```

**Comment by eed3si9n** [comment]
@olafurpg If the aim is to make `user.dir` consistent with the working directory in a forked run / test, it might make sense to append `-Duser.dir=...` at wherever that eventually does the forking like https://github.com/sbt/sbt/blob/v1.3.3/run/src/main/scala/sbt/Fork.scala based on the working directory that gets used.

**Comment by jozanek** [comment]
Status in sbt 2.x (verified on 2.0.8): since 2.0.4, forked `run` and `console` start in sbt's working directory, consistent with non-forked execution (#9442). Since 2.0.5, `Compile / run / baseDirectory := baseDirectory.value` restores the per-project directory (#9531).

Forked `test` intentionally keeps the project's `baseDirectory` — too many existing test suites depend on it. To opt in for tests:

```scala
Test / forkOptions := Def.uncached(
  (Test / forkOptions).value.withWorkingDirectory(None)
)
```

@eed3si9n given the test-side behavior is intentional and the opt-in exists, can this be closed?
